src_c/rect.c

```diff
@@ -695,6 +695,147 @@ pg_rect_collidelistall(pgRectObject *self, PyObject *args)
     return ret;
 }
 
+static SDL_Rect *
+pgRect_FromObjectAndKeyFunc(PyObject *obj, PyObject *keyfunc)
+{
+    SDL_Rect temp;
+    if (keyfunc == NULL) {
+        SDL_Rect *ret = pgRect_FromObject(obj, &temp);
+        if (!ret) {
+            PyErr_SetString(PyExc_TypeError,
+                            "Sequence must contain rect or rect-like objects");
+        }
+        return ret;
+    }
+
+    PyObject *obj_with_rect = PyObject_CallFunctionObjArgs(keyfunc, obj, NULL);
+    if (obj_with_rect == NULL) {
+        return NULL;
+    }
+
+    SDL_Rect *ret = pgRect_FromObject(obj_with_rect, &temp);
```

Comment:
ah, right, thanks

Reply:
So in general I can assume that when I get a pointer then the reference count has been increased?

Reply:
In general, functions that return python objects return "strong references", which means that it is up to the caller to `DECREF` a reference when the caller does not need it anymore. 

Functions that accept a python object as an argument generally "borrow references". This means that the caller function "holds a reference to the object", therefore the object should remain valid and accessible for the entire duration the subfunction is being called, which does not need to `DECREF` the borrowed reference because it does not "own that reference". In contrast to this, there can be some functions that "steal references" (usually not a recommended thing to do, but some python API does this). These are the kind of functions where a subfunction would `DECREF` its python argument, or put it to use elsewhere without `INCREF`ing it. This means that the caller essentially has "lost a reference" and now it's not guaranteed the object will remain a valid object after the function call. This can be easy worked around, the caller can `INCREF` an object before passing it to a function stealing a reference, so that the caller too holds a reference which is valid after the function call, if the caller needs to do anything with it later

When I refer to a "caller" anywhere above, I mean a function that is calling another function

Reply:
also about this line `Py_DECREF(obj_with_rect);`, it should be called regardless the error state, so this line needs moving to ouside the `if (!ret)` statement :)

Reply:
Thanks for the explanation. I know the concept of referencing, But to me it was not clear, where or which object is in control of the reference. I think you summed it up pretty well I guess it just takes a bit of practice or reading the documentation/source to understand who owns which reference.

Reply:
updated as suggested